File: frontend/src/services/aiProvider.ts
export type ProviderDescriptor = {
  id: 'openai' | 'google' | 'local_ollama' | 'local_lmstudio' | string
  label: string
  kind: 'remote' | 'local' | string
  requires_api_key: boolean
  supports_base_url: boolean
  default_base_url?: string | null
  supports_model_discovery: boolean
  supports_test: boolean
}

export type AiProviderKey = {
  id: number
  provider: string
  key_name: string
  key_masked: string
  created_at: string
  updated_at: string
}

export type AiActiveConfig = {
  enabled: boolean
  provider: string
  model?: string | null
  base_url?: string | null
  active_key_id?: number | null
  do_not_send_pii: boolean
  limits: {
    max_tokens_per_request?: number | null
    max_cost_usd_per_request?: number | null
    max_cost_usd_per_day?: number | null
  }
  active_key?: AiProviderKey | null
}

export type AiActiveConfigUpdate = Partial<{
  enabled: boolean
  provider: string
  model: string | null
  base_url: string | null
  active_key_id: number | null
  do_not_send_pii: boolean
  limits: AiActiveConfig['limits']
}>

export type ModelEntry = {
  id: string
  label: string
  source: 'discovered' | 'curated' | string
}

export async function fetchAiProviders(): Promise<ProviderDescriptor[]> {
  const res = await fetch('/api/ai/providers')
  if (!res.ok) throw new Error(`Failed to load AI providers (${res.status})`)
  return (await res.json()) as ProviderDescriptor[]
}

export async function fetchAiConfig(): Promise<AiActiveConfig> {
  const res = await fetch('/api/ai/config')
  if (!res.ok) throw new Error(`Failed to load AI config (${res.status})`)
  return (await res.json()) as AiActiveConfig
}

export async function updateAiConfig(payload: AiActiveConfigUpdate): Promise<AiActiveConfig> {
  const res = await fetch('/api/ai/config', {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`Failed to update AI config (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as AiActiveConfig
}

export async function listAiKeys(provider: string): Promise<AiProviderKey[]> {
  const url = new URL('/api/ai/keys', window.location.origin)
  url.searchParams.set('provider', provider)
  const res = await fetch(url.toString())
  if (!res.ok) throw new Error(`Failed to load keys (${res.status})`)
  return (await res.json()) as AiProviderKey[]
}

export async function createAiKey(payload: {
  provider: string
  key_name: string
  api_key_value: string
  meta?: Record<string, unknown>
}): Promise<AiProviderKey> {
  const res = await fetch('/api/ai/keys', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      provider: payload.provider,
      key_name: payload.key_name,
      api_key_value: payload.api_key_value,
      meta: payload.meta ?? {},
    }),
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`Failed to create key (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as AiProviderKey
}

export async function updateAiKey(
  id: number,
  payload: { key_name?: string; api_key_value?: string; meta?: Record<string, unknown> | null },
): Promise<AiProviderKey> {
  const res = await fetch(`/api/ai/keys/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`Failed to update key (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as AiProviderKey
}

export async function deleteAiKey(id: number): Promise<void> {
  const res = await fetch(`/api/ai/keys/${id}`, { method: 'DELETE' })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`Failed to delete key (${res.status})${body ? `: ${body}` : ''}`)
  }
}

export async function discoverAiModels(payload: {
  provider: string
  base_url?: string | null
  key_id?: number | null
}): Promise<ModelEntry[]> {
  const res = await fetch('/api/ai/models/discover', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      provider: payload.provider,
      base_url: payload.base_url ?? null,
      key_id: payload.key_id ?? null,
    }),
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`Failed to discover models (${res.status})${body ? `: ${body}` : ''}`)
  }
  const data = (await res.json()) as { models: ModelEntry[] }
  return data.models ?? []
}

export async function runAiTest(payload: {
  provider: string
  model: string
  base_url?: string | null
  key_id?: number | null
  prompt: string
}): Promise<{ text: string; latency_ms: number; usage: any; raw_metadata: any }> {
  const res = await fetch('/api/ai/test', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      provider: payload.provider,
      model: payload.model,
      base_url: payload.base_url ?? null,
      key_id: payload.key_id ?? null,
      prompt: payload.prompt,
    }),
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`AI test failed (${res.status})${body ? `: ${body}` : ''}`)
  }
  return (await res.json()) as { text: string; latency_ms: number; usage: any; raw_metadata: any }
}

File: backend/app/schemas/ai_provider.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ProviderDescriptor(BaseModel):
    id: str
    label: str
    kind: str  # remote|local
    requires_api_key: bool = False
    supports_base_url: bool = False
    default_base_url: Optional[str] = None
    supports_model_discovery: bool = True
    supports_test: bool = True


class AiProviderKeyRead(BaseModel):
    id: int
    provider: str
    key_name: str
    key_masked: str
    created_at: datetime
    updated_at: datetime


class AiProviderKeyCreate(BaseModel):
    provider: str
    key_name: str = Field(min_length=1, max_length=64)
    api_key_value: str = Field(min_length=1, max_length=4096)
    meta: Dict[str, Any] = Field(default_factory=dict)


class AiProviderKeyUpdate(BaseModel):
    key_name: Optional[str] = Field(default=None, min_length=1, max_length=64)
    api_key_value: Optional[str] = Field(default=None, min_length=1, max_length=4096)
    meta: Optional[Dict[str, Any]] = None


class AiLimits(BaseModel):
    max_tokens_per_request: Optional[int] = Field(default=None, ge=1)
    max_cost_usd_per_request: Optional[float] = Field(default=None, ge=0)
    max_cost_usd_per_day: Optional[float] = Field(default=None, ge=0)


class AiActiveConfig(BaseModel):
    enabled: bool = False
    provider: str = "openai"
    model: Optional[str] = None
    base_url: Optional[str] = None
    active_key_id: Optional[int] = None
    do_not_send_pii: bool = True
    limits: AiLimits = Field(default_factory=AiLimits)

    # Optional denormalized key metadata for UI.
    active_key: Optional[AiProviderKeyRead] = None


class AiActiveConfigUpdate(BaseModel):
    enabled: Optional[bool] = None
    provider: Optional[str] = None
    model: Optional[str] = None
    base_url: Optional[str] = None
    active_key_id: Optional[int] = None
    do_not_send_pii: Optional[bool] = None
    limits: Optional[AiLimits] = None


class DiscoverModelsRequest(BaseModel):
    provider: str
    base_url: Optional[str] = None
    key_id: Optional[int] = None


class ModelEntry(BaseModel):
    id: str
    label: str
    source: str = "discovered"  # discovered|curated
    raw: Dict[str, Any] = Field(default_factory=dict)


class DiscoverModelsResponse(BaseModel):
    models: List[ModelEntry] = Field(default_factory=list)


class AiTestRequest(BaseModel):
    provider: Optional[str] = None
    model: Optional[str] = None
    base_url: Optional[str] = None
    key_id: Optional[int] = None
    prompt: str = Field(min_length=1, max_length=4000)


class AiUsage(BaseModel):
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    total_tokens: Optional[int] = None


class AiTestResponse(BaseModel):
    text: str
    latency_ms: int
    usage: AiUsage = Field(default_factory=AiUsage)
    raw_metadata: Dict[str, Any] = Field(default_factory=dict)


__all__ = [
    "AiActiveConfig",
    "AiActiveConfigUpdate",
    "AiProviderKeyCreate",
    "AiProviderKeyRead",
    "AiProviderKeyUpdate",
    "AiTestRequest",
    "AiTestResponse",
    "DiscoverModelsRequest",
    "DiscoverModelsResponse",
    "ModelEntry",
    "ProviderDescriptor",
]

